odl/space/fspace.py

```diff
@@ -25,51 +25,254 @@
+    return out
+
+
+def vectorize(dtype, outarg='none'):
+    """Vectorization decorator for our input parameter pattern.
+
+    The wrapped function must be callable with one positional
+    parameter. Keyword arguments are passed through, hence positional
+    arguments with defaults can either be left out or passed by keyword,
+    but not by position.
+
+    Parameters
+    ----------
+    dtype : object
+        Data type of the output array. Needs to be understood by the
+        `numpy.dtype` function.
+    outarg : {'none', 'positional', 'optional'}
+        Type of the output argument of the decorated function for
+        in-place evaluation
+
+        'none': No output parameter. This is the default.
+        Resulting argspec: ``func(x, **kwargs)``
+        Returns: the new array
+
+        'positional': Required argument ``out`` at second position.
+        Resulting argspec: ``func(x, out=None, **kwargs)``
+        Returns: ``out``
+
+        'optional': optional argument ``out`` with default `None`.
+        Resulting argspec: ``func(x, out=None, **kwargs)``
+        Returns: ``out`` if it is not `None` otherwise a new array
+
+    Note
+    ----
+    For ``outarg`` not equal to 'none', the decorated function returns
+    the array given as ``out`` argument if it is not `None`.
+
+    Examples
+    --------
+    Vectorize a step function in the first variable:
+
+    >>> @vectorize(dtype=float)
+    ... def step(x):
+    ...     return 0 if x[0] <= 0 else 1
+
+    This corresponds to (but is much slower than)
+
+    >>> import numpy as np
+    >>> def step_vec(x):
+    ...     x0, x1 = x
+    ...     # np.broadcast is your friend to determine the output shape
+    ...     out = np.zeros(np.broadcast(x0, x1).shape, dtype=x0.dtype)
+    ...     idcs = np.where(x0 > 0)
+    ...     # Need to throw away the indices from the empty dimensions
+    ...     idcs = idcs[0] if len(idcs) > 1 else idcs
+    ...     out[idcs] = 1
+    ...     return out
+
+    Both versions work for arrays and meshgrids:
+
+    >>> x = np.linspace(-5, 13, 10, dtype=float).reshape((2, 5))
+    >>> x  # array representing 5 points in 2d
+    array([[ -5.,  -3.,  -1.,   1.,   3.],
+           [  5.,   7.,   9.,  11.,  13.]])
+    >>> np.array_equal(step(x), step_vec(x))
+    True
+
+    >>> x = y = np.linspace(-1, 2, 5)
+    >>> mg_sparse = np.meshgrid(x, y, indexing='ij', sparse=True)
+    >>> np.array_equal(step(mg_sparse), step_vec(mg_sparse))
+    True
+    >>> mg_dense = np.meshgrid(x, y, indexing='ij', sparse=False)
+    >>> np.array_equal(step(mg_dense), step_vec(mg_dense))
+    True
+
+    With output parameter:
+
+    >>> @vectorize(dtype=float, outarg='positional')
+    ... def step(x):
+    ...     return 0 if x[0] <= 0 else 1
+    >>> x = np.linspace(-5, 13, 10, dtype=float).reshape((2, 5))
+    >>> out = np.empty(5, dtype=float)
+    >>> step(x, out)  # returns out
+    array([ 0.,  0.,  0.,  1.,  1.])
+    """
+    def vect_decorator(func):
```

Comment:
Very complicated. Warning bells ring. Do we test all paths? Perhaps split in smaller somehow?